File: Sources/ThesisSkipApp/ui/pages/reservations_page/partials/ReservationDetail.swift
//
//  ReservationDetail.swift
//  MADiOS
//
//  Created by Bram Rampelberg on 02/01/2025.
//  Copyright © 2025 HOGENT. All rights reserved.
//

import SwiftUI

struct ReservationDetail: View {
    @Environment(\.verticalSizeClass) private var verticalSizeClass
    @EnvironmentObject var reservationViewModel: ReservationsViewModel
    let reservation: Reservation
    let reservationDetails: ReservationDetails?
    let isReservationCancelable: Bool
    
    @State private var showCancelErrorAlert = false
    
    var body: some View {
        Form {
            Section(header: Text("Reservation Details").font(.title2)) {
                ImportantReservationInfo(reservation: reservation)
            }
            
            if reservationDetailsAreValid() {
                Section(header: Text("Acceptee Info")) {
                    infoRow(label: "Name", value: reservationDetails!.currentBatteryUserName!)
                    infoRow(label: "Phone", value: reservationDetails!.currentHolderPhoneNumber!)
                    infoRow(label: "E-mail", value: reservationDetails!.currentHolderEmail!)
                }
                
                Section(header: Text("Address")) {
                    infoRow(label: "Street", value: "\(reservationDetails!.currentHolderStreet!) \(reservationDetails!.currentHolderNumber!)")
                    infoRow(label: "City", value: "\(reservationDetails!.currentHolderPostalCode!) \(reservationDetails!.currentHolderCity!)")
                }
                
                if let mentorName = reservationDetails?.mentorName, !mentorName.isEmpty {
                    Section(header: Text("Mentor")) {
                        Text(mentorName)
                    }
                }
            } else {
                Section {
                    detailsNotAvailable
                }
            }
            
            if reservationViewModel.isCancelationPending {
                Section {
                    HStack {
                        Spacer()
                        ProgressView().padding()
                        Spacer()
                    }
                }
            }
            
            if isReservationCancelable {
                Section {
                    cancelButton
                }
            }
        }
//        .presentationDetents([
//            reservationDetailsAreValid() ?
//            UIDevice.current.userInterfaceIdiom == .pad ? .fraction(0.77) : .fraction(0.7)
//            :
//                UIDevice.current.userInterfaceIdiom == .pad ? .medium : .fraction(0.5),
//            .large
//        ])
    }
    
    func infoRow(label: String, value: String) -> some View {
        HStack {
            Text(label)
                .foregroundColor(.secondary)
            Spacer()
            Text(value)
                .multilineTextAlignment(.trailing)
        }
    }
    
    var detailsNotAvailable: some View {
        VStack {
            Image(systemName: "info.circle.fill")
                .resizable()
                .frame(width: 48, height: 48)
                .foregroundColor(.blue)
                .padding(.top, 20)
            
            Text("No pickup information available")
                .font(.body)
                .multilineTextAlignment(.center)
                .padding(.top, 8)
        }
        .frame(maxWidth: .infinity)
    }
    
    var cancelButton: some View {
        Button(role: .destructive, action: {reservationViewModel.cancelReservation()}) {
            Label("Cancel Reservation", systemImage: "xmark.circle.fill")
                .frame(maxWidth: .infinity)
                .padding()
                .foregroundColor(.red)
        }
    }
    
    func reservationDetailsAreValid() -> Bool {
        return !(
            reservationDetails?.currentBatteryUserName?.isEmpty ?? true ||
            reservationDetails?.currentHolderPhoneNumber?.isEmpty ?? true ||
            reservationDetails?.currentHolderEmail?.isEmpty ?? true ||
            reservationDetails?.currentHolderStreet?.isEmpty ?? true ||
            reservationDetails?.currentHolderNumber?.isEmpty ?? true ||
            reservationDetails?.currentHolderCity?.isEmpty ?? true ||
            reservationDetails?.currentHolderPostalCode?.isEmpty ?? true
        )
    }
}

#Preview {
    @Previewable var reservationViewModel = ReservationsViewModel()
    
    ReservationDetail(
        reservation: Reservation(start: Date(), end: Date(), date: Date(), boatId: 1, boatPersonalName: "boatName", id: 0, isDeleted: false),
        reservationDetails: ReservationDetails(mentorName: "mentor", batteryId: 1, currentBatteryUserName: "username", currentBatteryUserId: 1, currentHolderPhoneNumber: "phonenumber", currentHolderEmail: "email", currentHolderStreet: "street", currentHolderNumber: "number", currentHolderCity: "city", currentHolderPostalCode: "postalCode"),
        isReservationCancelable: true
    ).environmentObject(reservationViewModel)
}

File: .build/plugins/outputs/thesis-skip-app/ThesisSkipApp/destination/skipstone/ThesisSkipApp/src/main/kotlin/thesis/skip/app/ReservationDetail.kt
//
//  ReservationDetail.swift
//  MADiOS
//
//  Created by Bram Rampelberg on 02/01/2025.
//  Copyright © 2025 HOGENT. All rights reserved.
//

package thesis.skip.app

import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.saveable.Saver
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import skip.lib.*
import skip.lib.Set

import skip.ui.*
import skip.foundation.*
import skip.model.*

internal class ReservationDetail: View {
    private var verticalSizeClass: UserInterfaceSizeClass? = null
    internal var reservationViewModel: ReservationsViewModel
        get() = _reservationViewModel.wrappedValue
        set(newValue) {
            _reservationViewModel.wrappedValue = newValue
        }
    internal var _reservationViewModel = skip.ui.Environment<ReservationsViewModel>()
    internal val reservation: Reservation
    internal val reservationDetails: ReservationDetails?
    internal val isReservationCancelable: Boolean

    private var showCancelErrorAlert: Boolean
        get() = _showCancelErrorAlert.wrappedValue
        set(newValue) {
            _showCancelErrorAlert.wrappedValue = newValue
        }
    private var _showCancelErrorAlert: skip.ui.State<Boolean>

    override fun body(): View {
        return ComposeBuilder { composectx: ComposeContext ->
            Form { ->
                ComposeBuilder { composectx: ComposeContext ->
                    Section(header = Text(LocalizedStringKey(stringLiteral = "Reservation Details")).font(Font.title2)) { ->
                        ComposeBuilder { composectx: ComposeContext ->
                            ImportantReservationInfo(reservation = reservation).Compose(composectx)
                            ComposeResult.ok
                        }
                    }.Compose(composectx)

                    if (reservationDetailsAreValid()) {
                        Section(header = Text(LocalizedStringKey(stringLiteral = "Acceptee Info"))) { ->
                            ComposeBuilder { composectx: ComposeContext ->
                                infoRow(label = "Name", value = reservationDetails!!.currentBatteryUserName!!).Compose(composectx)
                                infoRow(label = "Phone", value = reservationDetails!!.currentHolderPhoneNumber!!).Compose(composectx)
                                infoRow(label = "E-mail", value = reservationDetails!!.currentHolderEmail!!).Compose(composectx)
                                ComposeResult.ok
                            }
                        }.Compose(composectx)

                        Section(header = Text(LocalizedStringKey(stringLiteral = "Address"))) { ->
                            ComposeBuilder { composectx: ComposeContext ->
                                infoRow(label = "Street", value = "${reservationDetails!!.currentHolderStreet!!} ${reservationDetails!!.currentHolderNumber!!}").Compose(composectx)
                                infoRow(label = "City", value = "${reservationDetails!!.currentHolderPostalCode!!} ${reservationDetails!!.currentHolderCity!!}").Compose(composectx)
                                ComposeResult.ok
                            }
                        }.Compose(composectx)

                        reservationDetails?.mentorName?.let { mentorName ->
                            if (!mentorName.isEmpty) {
                                Section(header = Text(LocalizedStringKey(stringLiteral = "Mentor"))) { ->
                                    ComposeBuilder { composectx: ComposeContext ->
                                        Text(mentorName).Compose(composectx)
                                        ComposeResult.ok
                                    }
                                }.Compose(composectx)
                            }
                        }
                    } else {
                        Section { ->
                            ComposeBuilder { composectx: ComposeContext ->
                                detailsNotAvailable.Compose(composectx)
                                ComposeResult.ok
                            }
                        }.Compose(composectx)
                    }

                    if (reservationViewModel.isCancelationPending) {
                        Section { ->
                            ComposeBuilder { composectx: ComposeContext ->
                                HStack { ->
                                    ComposeBuilder { composectx: ComposeContext ->
                                        Spacer().Compose(composectx)
                                        ProgressView().padding().Compose(composectx)
                                        Spacer().Compose(composectx)
                                        ComposeResult.ok
                                    }
                                }.Compose(composectx)
                                ComposeResult.ok
                            }
                        }.Compose(composectx)
                    }

                    if (isReservationCancelable) {
                        Section { ->
                            ComposeBuilder { composectx: ComposeContext ->
                                cancelButton.Compose(composectx)
                                ComposeResult.ok
                            }
                        }.Compose(composectx)
                    }
                    ComposeResult.ok
                }
            }
            .presentationDetents(setOf(
                if (reservationDetailsAreValid()) PresentationDetent.fraction(0.7) else PresentationDetent.fraction(0.5),
                PresentationDetent.large
            )).Compose(composectx)
        }
    }

    @Composable
    @Suppress("UNCHECKED_CAST")
    override fun Evaluate(context: ComposeContext, options: Int): kotlin.collections.List<Renderable> {
        val rememberedshowCancelErrorAlert by rememberSaveable(stateSaver = context.stateSaver as Saver<skip.ui.State<Boolean>, Any>) { mutableStateOf(_showCancelErrorAlert) }
        _showCancelErrorAlert = rememberedshowCancelErrorAlert

        this.verticalSizeClass = EnvironmentValues.shared.verticalSizeClass
        _reservationViewModel.wrappedValue = EnvironmentValues.shared.environmentObject(type = ReservationsViewModel::class)!!

        return super.Evaluate(context, options)
    }

    internal fun infoRow(label: String, value: String): View {
        return HStack { ->
            ComposeBuilder { composectx: ComposeContext ->
                Text(label)
                    .foregroundColor(Color.secondary).Compose(composectx)
                Spacer().Compose(composectx)
                Text(value)
                    .multilineTextAlignment(TextAlignment.trailing).Compose(composectx)
                ComposeResult.ok
            }
        }
    }

    internal val detailsNotAvailable: View
        get() {
            return VStack { ->
                ComposeBuilder { composectx: ComposeContext ->
                    Image(systemName = "info.circle.fill")
                        .resizable()
                        .frame(width = 48.0, height = 48.0)
                        .foregroundColor(Color.blue)
                        .padding(Edge.Set.top, 20.0).Compose(composectx)

                    Text(LocalizedStringKey(stringLiteral = "No pickup information available"))
                        .font(Font.body)
                        .multilineTextAlignment(TextAlignment.center)
                        .padding(Edge.Set.top, 8.0).Compose(composectx)
                    ComposeResult.ok
                }
            }
            .frame(maxWidth = Double.infinity)
        }

    internal val cancelButton: View
        get() {
            return Button(role = ButtonRole.destructive, action = { -> reservationViewModel.cancelReservation() }) { ->
                ComposeBuilder { composectx: ComposeContext ->
                    Label(LocalizedStringKey(stringLiteral = "Cancel Reservation"), systemImage = "xmark.circle.fill")
                        .frame(maxWidth = Double.infinity)
                        .padding()
                        .foregroundColor(Color.red).Compose(composectx)
                    ComposeResult.ok
                }
            }
        }

    internal fun reservationDetailsAreValid(): Boolean {
        return !(reservationDetails?.currentBatteryUserName?.isEmpty ?: true || reservationDetails?.currentHolderPhoneNumber?.isEmpty ?: true || reservationDetails?.currentHolderEmail?.isEmpty ?: true || reservationDetails?.currentHolderStreet?.isEmpty ?: true || reservationDetails?.currentHolderNumber?.isEmpty ?: true || reservationDetails?.currentHolderCity?.isEmpty ?: true || reservationDetails?.currentHolderPostalCode?.isEmpty ?: true)
    }

    private constructor(reservation: Reservation, reservationDetails: ReservationDetails? = null, isReservationCancelable: Boolean, showCancelErrorAlert: Boolean = false, privatep: Nothing? = null) {
        this.reservation = reservation.sref()
        this.reservationDetails = reservationDetails.sref()
        this.isReservationCancelable = isReservationCancelable
        this._showCancelErrorAlert = skip.ui.State(showCancelErrorAlert)
    }

    constructor(reservation: Reservation, reservationDetails: ReservationDetails? = null, isReservationCancelable: Boolean): this(reservation = reservation, reservationDetails = reservationDetails, isReservationCancelable = isReservationCancelable, privatep = null) {
    }
}

// #Preview omitted
